src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokensFileViewProvider.java

```diff
@@ -0,0 +1,246 @@
+ * Copyright (c) 2025 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.semanticTokens.viewProvider;
+
+import com.intellij.lang.Language;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.*;
+import com.intellij.psi.util.CachedValueProvider;
+import com.intellij.psi.util.CachedValuesManager;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.client.features.EditorBehaviorFeature;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link FileViewProvider} for an LSP-backed TextMate file where elements are derived dynamically from reported
+ * semantic tokens.
+ */
+@ApiStatus.Internal
+public class LSPSemanticTokensFileViewProvider extends SingleRootFileViewProvider {
+
+    /**
+     * Creates a new file view provider.
+     *
+     * @param psiManager         the PSI manager
+     * @param virtualFile        the virtual file
+     * @param eventSystemEnabled whether or not the event system is enabled
+     * @param language           the language which should always be TextMate
+     */
+    LSPSemanticTokensFileViewProvider(@NotNull PsiManager psiManager,
+                                      @NotNull VirtualFile virtualFile,
+                                      boolean eventSystemEnabled,
+                                      @NotNull Language language) {
+        super(psiManager, virtualFile, eventSystemEnabled, language);
+    }
+
+    /**
+     * Returns the semantic tokens file view provider for the provided PSI file if assignable and enabled.
+     *
+     * @param file the PSI file
+     * @return the semantic tokens file view provider if assignable and enabled; otherwise false
+     */
+    @ApiStatus.Internal
+    public static LSPSemanticTokensFileViewProvider getInstance(@Nullable PsiFile file) {
+        FileViewProvider fileViewProvider = file != null ? file.getViewProvider() : null;
+        return ((fileViewProvider instanceof LSPSemanticTokensFileViewProvider semanticTokensFileViewProvider) &&
+                semanticTokensFileViewProvider.isEnabled()) ?
+                semanticTokensFileViewProvider :
+                null;
+    }
+
+    // Only override standard behavior when enabled to do so
+    private boolean isEnabled() {
```

Comment:
This and `getFile()` below are how the feature flag is used. All overrides of inherited behavior _must_ consult `isEnabled()` and delegate to the default/inherited behavior if not enabled. That allows immediate switching of behavior when the feature flag is toggled, though due to PSI caching, the full effect may not be seen until after the PSI tree has changed.